Source/Dafny/Rewriter.cs

```diff
@@ -118,14 +120,58 @@ public abstract class IRewriter {
       Contract.Requires(program != null);
     }
   }
+  /*
+   * Code Example that displays a Warning
+   * 
+   * datatype Color = Red | Green | ShadesOfGray(nat)
+      method MonochromaticMethod(c: Color) returns (x: bool) {
+      return match c
+            case ShadesOfGray => true
+            case anythingElse => false;
+    }
+   */
+  class ConstructorMatchWarning : IRewriter {
+    internal override void PostResolve(Program program) {
+      base.PostResolve(program);
+      foreach (var moduleDefinition in program.Modules()) {
+        foreach (var topLevelDecl in moduleDefinition.TopLevelDecls.OfType<TopLevelDeclWithMembers>()) {
+          foreach (var callable in topLevelDecl.Members.OfType<ICallable>()) {
+            var visitor = new ConstructorMatchWarningVisitor(this.Reporter);
+            visitor.Visit(callable, Unit.Default);
+          }
+        }
+      }
+    }
+    public ConstructorMatchWarning(ErrorReporter reporter) : base(reporter) {
+    }
+  }
 
+  class ConstructorMatchWarningVisitor : TopDownVisitor<Unit> {
+    private readonly ErrorReporter reporter;
+    public ConstructorMatchWarningVisitor(ErrorReporter reporter) {
+      this.reporter = reporter;
+    }
+
+    protected override bool VisitOneExpr(Expression expr, ref Unit st) {
+      if (expr is NestedMatchExpr matchExpr) {
+        var idPatterns = matchExpr.Cases.Select(matchCase => matchCase.Pat).OfType<IdPattern>();
+        foreach (var idPattern in idPatterns) {
+          var variableChecker = idPattern.Arguments;
```

Comment:
I don't understand the name `variableChecker`. With my previous comment I was thinking of something like 
`var isVariable = idPattern.Argument == null`